sanic/protocols/http2.py

```diff
@@ -0,0 +1,134 @@
+import asyncio
+import io
+import collections
+from functools import partial
+from typing import List, Tuple
+
+from h2.connection import H2Connection
+from h2.events import (
+    ConnectionTerminated, DataReceived, RequestReceived, StreamEnded
+)
+from h2.errors import ErrorCodes
+from h2.exceptions import ProtocolError
+
+from sanic.request import Request
+from sanic.log import log
+
+RequestData = collections.namedtuple('RequestData', ['headers', 'data'])
+
+
+def generate_http2_protocol(ssl, parent_class):
+    ssl.set_ciphers("ECDHE+AESGCM")
+    ssl.set_alpn_protocols(["h2", "http/1.1"])
+
+    _HTTP2Protocol = type("HTTP2Protocol", (HTTP2Checker, parent_class), {})
+    _HTTP2Protocol.protocol_class = type("H2Protocol",
+                                         (Http2Protocol, parent_class), {})
+
+    return _HTTP2Protocol
+
+
+class HTTP2Checker:
+    def connection_made(self, transport):
+        ssl_object = transport.get_extra_info('ssl_object')
+        protocol = ssl_object.selected_alpn_protocol()
+
+        if protocol == 'h2':
+            self.__class__ = self.protocol_class
+            self.connection_made(transport)
+        else:
+            super().connection_made(transport)
+
+
+class Http2Protocol(asyncio.Protocol):
+
+    def connection_made(self, transport):
+        # TODO: Request timeouts
+        self.h2_connection = H2Connection(client_side=False)
+        self.stream_data = {}
+        self.h2_connection.initiate_connection()
+        self.transport = transport
+        self.transport.write(self.h2_connection.data_to_send())
+
+    def data_received(self, data: bytes):
+        try:
+            events = self.h2_connection.receive_data(data)
+        except ProtocolError as e:
+            self.transport.write(self.h2_connection.data_to_send())
+            self.transport.close()
+        else:
+            self.transport.write(self.h2_connection.data_to_send())
+            for event in events:
+                if isinstance(event, RequestReceived):
+                    self.receive_headers(event.headers, event.stream_id)
+                elif isinstance(event, DataReceived):
+                    self.receive_data(event.data, event.stream_id)
+                elif isinstance(event, StreamEnded):
+                    self.stream_complete(event.stream_id)
+                elif isinstance(event, ConnectionTerminated):
+                    self.transport.close()
```

Comment:
This code doesn't appear to handle flow control. It's definitely the most subtle part of the whole procedure, and you'll need it to get the server to be really stable. I appreciate this is an early draft of the patch so it's a natural thing to leave out, but for when you get around to it I wrote some [introductory documentation](https://python-hyper.org/h2/en/stable/advanced-usage.html#flow-control) that might help.